openrtb_ext/request.go

```diff
@@ -71,14 +71,155 @@ func (ert *ExtRequestTargeting) UnmarshalJSON(b []byte) error {
 }
 
 // PriceGranularity defines the allowed values for bidrequest.ext.prebid.targeting.pricegranularity
-type PriceGranularity string
-
-const (
-	PriceGranularityLow    PriceGranularity = "low"
-	PriceGranularityMedium PriceGranularity = "medium"
-	// Seems that PBS was written with medium = "med", so hacking that in
-	PriceGranularityMedPBS PriceGranularity = "med"
-	PriceGranularityHigh   PriceGranularity = "high"
-	PriceGranularityAuto   PriceGranularity = "auto"
-	PriceGranularityDense  PriceGranularity = "dense"
-)
+type PriceGranularity struct {
+	Precision int                `json:"precision,omitempty"`
+	Ranges    []GranularityRange `json:"ranges,omitempty"`
+}
+
+type PriceGranularityRaw PriceGranularity
+
+// GranularityRange struct defines a range of prices used by PriceGranularity
+type GranularityRange struct {
+	Min       float64 `json:"min"`
+	Max       float64 `json:"max"`
+	Increment float64 `json:"increment"`
+}
+
+// UnmarshalJSON : custom unmarshaller to handle legacy string granularites.
+func (pg *PriceGranularity) UnmarshalJSON(b []byte) error {
+	// We default to medium
+	if len(b) == 0 {
+		*pg = priceGranularityMed
+		return nil
+	}
+	// First check for legacy strings
+	var pgString string
+	err := json.Unmarshal(b, &pgString)
+	if err == nil {
+		*pg = PriceGranularityFromString(pgString)
+		if len(pg.Ranges) > 0 {
+			// Only exit if we matched something, else we try processing as custom granularity
+			// This way we error as expecting the new custom granularity standard.
+			return nil
+		}
+	}
+	// Not legacy, so we do a normal Unmarshal
+	pgraw := PriceGranularityRaw{}
+	pgraw.Precision = 2
+	err = json.Unmarshal(b, &pgraw)
+	if err != nil {
+		return err
+	}
+	if pgraw.Precision < 0 {
+		return errors.New("Price granularity error: precision must be non-negative")
+	}
+	if len(pgraw.Ranges) > 0 {
+		var prevMax float64 = 0
+		for i, gr := range pgraw.Ranges {
+			if gr.Max < prevMax {
+				return errors.New("Price granularity error: range list must be ordered with increasing \"max\"")
+			}
+			if gr.Max < gr.Min {
+				return errors.New("Price granularity error: max must be greater than min")
+			}
+			if gr.Increment <= 0.0 {
+				return errors.New("Price granularity error: increment must be a nonzero positive number")
+			}
+			// Enforce that we don't read "min" from the request
+			pgraw.Ranges[i].Min = prevMax
+			if pgraw.Ranges[i].Min < prevMax {
+				return errors.New("Price granularity error: overlapping granularity ranges")
+			}
+			prevMax = gr.Max
+		}
+	} else if len(pgraw.Ranges) == 0 {
```

Comment:
This could just be a simple `else`... if `len(arr)` isn't positive, it'd be 0.

Reply:
Yes, a hold over from when I only ran checks when `len()>1`